Fix incorrect window start position

ZarsBranchkin opened the pull request:
If the value of `scroll-margin` is not 0, the live buffer won't align with source code buffer properly.

Wasn't sure whether to compensate for `scroll-margin` manually, so decided to set it to 0 locally for sake of simplicity.

donkirkby commented:
Thanks for the commit, I didn't know about `scroll-margin`. I can reproduce the problem, but the commit doesn't seem to fix it. Maybe the problem I found isn't the same as yours. Please post your steps to reproduce the problem, and tell me whether your commit works for you with the problem described here.

Here are the steps I used to reproduce the problem:
1. Open a python file: `emacs foo.py`
2. Type some Python code that is more than one screen.
3. `M-x live-py-mode`
4. `M-x set-variable` `scroll-margin` `3`
5. Scroll off the screen.

Expected result: display stays lined up with the code.

Actual result: display is off by 3 lines.

brandm commented on emacs-live-py-mode/live-py-mode.el:
```diff
@@ -147,6 +147,7 @@ When the source buffer is narrowed the trace buffer remains
 aligned but will not hide the part after the narrowing."
   (let* ((output-window (get-buffer-window live-py-trace-name))
          (point-min-pos (point-min))
+         (scroll-margin 0)
```
I suggest

  `(let ((scroll-margin 0)) (recenter-top-bottom 0))`

instead to make it clear where the setting is needed.

brandm commented:
> 1. Open a python file: `emacs foo.py`
> 2. Type some Python code that is more than one screen.
> 3. `M-x live-py-mode`
> 4. `M-x set-variable` `scroll-margin` `3`
> 5. Scroll off the screen.

With Emacs 25.2 and 25.1 (tried no older) I can reproduce the issue and the fix.

ZarsBranchkin commented:
@donkirkby Yeah, that is the same issue. After tinkering around some more, I managed to break the commit as well.

It seems that instead of using `let`, the `scroll-margin` has to be set locally for the output buffer.

In the https://github.com/donkirkby/live-py-plugin/pull/127/commits/af9bd0b798d8ffb695b6739cba7f2ccdf6c481c9  commit I also simplified the update scroll function a bit.

As for the https://github.com/donkirkby/live-py-plugin/pull/127/commits/85aefe383196508ca362b5f6f6fe3c057584b2d9 commit, it's supposed to be an idea for aligning the source code and output buffer line count, to avoid misaligned code when the source has more lines than output and point is at the end of the buffer. Wasn't sure where to put the code though.

donkirkby commented:
That works for me, now. Thanks.

Adding the extra lines broke some of the emacs tests, do you want to fix them, or would you like me to do it? You can see the failures more clearly by running them interactively. There are instructions in the [test file], but I ran them like this:

    emacs -Q -nw -L emacs-live-py-mode -L plugin/PySrc -l live-py-mode.el -l live-py-test.el -f ert-run-tests-interactively

If you're not familiar with the test summary (I wasn't), the [docs] are helpful.

There are two more [broken tests] that have something to do with abandoning an open window. @brandm might be able to clarify why they're failing.

[docs]: http://www.gnu.org/software/emacs/manual/html_node/ert/Understanding-Explanations.html#Understanding-Explanations 
[test file]: https://github.com/donkirkby/live-py-plugin/blob/ee9c0e8ca21599bd2708e3cbda7fc097a119cc73/emacs-live-py-mode/live-py-test.el#L25
[broken tests]: https://github.com/donkirkby/live-py-plugin/blob/ee9c0e8ca21599bd2708e3cbda7fc097a119cc73/emacs-live-py-mode/live-py-test.el#L329

brandm commented on emacs-live-py-mode/live-py-mode.el:
```diff
@@ -146,22 +146,17 @@ both are relative to (point-min). Numbering starts at 1 for all
-                              (save-restriction
-                                (widen)
-                                (line-number-at-pos point-min-pos)))))
+         (point-min-line-nr (count-lines 1 (point-min)))
```
Nice solution to compensate for narrowing.

brandm commented on emacs-live-py-mode/live-py-mode.el:
```diff
@@ -146,22 +146,17 @@ both are relative to (point-min). Numbering starts at 1 for all
+      (forward-line window-start-line-nr)
       (recenter-top-bottom 0)
-      (forward-line (- point-line-nr window-start-line-nr)))
-    (set-window-buffer output-window live-py-trace-name)))
```
This move breaks

live-py-test/3-other-window/4-abandon-source-and-swap/edit
live-py-test/3-other-window/4-abandon-source-and-swap/move

What issue is the move expected to fix?

ZarsBranchkin replied:
Ah, alright, didn't know it was that important. Just seemed odd to display a buffer that was already checked as visible.

brandm commented on emacs-live-py-mode/live-py-mode.el:
```diff
@@ -170,7 +165,7 @@ aligned but will not hide the part after the narrowing."
   ;; to miss the error. And on such an error the function will be removed
   ;; from `post-command-hook' which is confusing when not noticed.
-  (when (memq this-command '(narrow-to-region next-line viper-goto-line))
+  (when (memq this-command '(narrow-to-region next-line previous-line viper-goto-line))
```
For performance reasons this list should be as short as possible because it requires a `redisplay` for reasons explained in the comments. Why add `previous-line`? If it is just for symmetry with `next-line` then please remove `previous-line` again and if it helps add a comment below the comment for `next-line` that it is not in the list intentionally.

brandm commented:
> Wasn't sure where to put the code though.

If the backend would add the extra lines for the line count alignment instead of the Emacs frontend then the other frontends would benefit as well.

brandm commented:
> Adding the extra lines broke some of the emacs tests,

Yes, the expected output buffer `live-py-test-trace` needs the line count aligned.

ZarsBranchkin commented:
Alright, fixed the mentioned issues and moved the line alignment to the live-py backend. 

I ran the tests and two of them failed due to the appended lines, but that's obviously intentional in this case.

donkirkby commented:
I fixed the broken unit tests by adding some extra blank lines.

Thanks for the fix, @ZarsBranchkin, and thanks for the review, @brandm.
